network/src/sync/master.rs

```diff
@@ -0,0 +1,304 @@
+// Copyright (C) 2019-2021 Aleo Systems Inc.
+// This file is part of the snarkOS library.
+
+// The snarkOS library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The snarkOS library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the snarkOS library. If not, see <https://www.gnu.org/licenses/>.
+
+use std::{collections::HashMap, net::SocketAddr, time::Duration};
+
+use crate::{NetworkError, Node, Payload, Peer};
+use futures::{pin_mut, select, FutureExt};
+use rand::prelude::SliceRandom;
+use snarkvm_algorithms::crh::double_sha256;
+use snarkvm_objects::{BlockHeader, BlockHeaderHash, Storage};
+use tokio::{sync::mpsc, time::Instant};
+
+pub enum SyncInbound {
+    BlockHashes(SocketAddr, Vec<BlockHeaderHash>),
+    Block(SocketAddr, Vec<u8>),
+}
+
+pub struct SyncMaster<S: Storage + Send + Sync + 'static> {
+    node: Node<S>,
+    incoming: mpsc::Receiver<SyncInbound>,
+}
+
+struct SyncBlock {
+    address: SocketAddr,
+    block: Vec<u8>,
+}
+
+impl<S: Storage + Send + Sync + 'static> SyncMaster<S> {
+    pub fn new(node: Node<S>) -> (Self, mpsc::Sender<SyncInbound>) {
+        let (sender, receiver) = mpsc::channel(256);
+        let new = Self {
+            node,
+            incoming: receiver,
+        };
+        (new, sender)
+    }
+
+    async fn find_sync_nodes(&mut self) -> Vec<Peer> {
+        let our_block_height = self.node.expect_sync().current_block_height();
+        let mut interesting_peers = vec![];
+        for node in self.node.peer_book.connected_peers_snapshot().await {
+            if !node.judge() && node.quality.block_height > our_block_height + 1 {
+                interesting_peers.push(node);
+            }
+        }
+        interesting_peers.sort_by(|x, y| y.quality.block_height.cmp(&x.quality.block_height));
```

Comment:
do we have a method to extract the peer's height? a `sort_by_key` would be more concise here

Reply:
perhaps paired with a `cmp::Reverse` too

Reply:
perhaps, this is more readable though imo